docs/devsite-help/grpc-streaming.md

```diff
@@ -0,0 +1,130 @@
+
+APIs using bidirectional-streaming calls vary significantly, and you should consult the
+API documentation for details. One common pattern is for the *first* client request to
+contain some initialization data (for example, the name of a data store to write to) and
+subsequent requests are slightly different. Again, consult the API-specific documentation
+for details.
+
+In the Google Cloud Libraries for .NET, an ongoing server-streaming call
+is represented by the base class `BidirectionalStreamingBase<TRequest, TResponse>`, with a
+separate derived class for each method. For example, in the client library for the
+[BigQuery Storage API](https://cloud.google.com/bigquery/docs/reference/storage),
+the [BigQueryWriteClient.AppenRows](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.BigQuery.Storage.V1/latest/Google.Cloud.BigQuery.Storage.V1.BigQueryWriteClient#Google_Cloud_BigQuery_Storage_V1_BigQueryWriteClient_AppendRows_Google_Api_Gax_Grpc_CallSettings_Google_Api_Gax_Grpc_BidirectionalStreamingSettings_) method returns a [BigQueryWriteClient.AppensRowsStream](https://cloud.google.com/dotnet/docs/reference/Google.Cloud.BigQuery.Storage.V1/latest/Google.Cloud.BigQuery.Storage.V1.BigQueryWriteClient.AppendRowsStream).
+
+`BidirectionalStreamingBase<TRequest, TResponse>` has all the aspects of `ServerStreamingBase<TResponse>` described above, but with additional methods related to the client's request stream:
+
+- `WriteAsync` and `TryWriteAsync` send a request
+- `CompleteAsync` and `TryCompleteAsync` are called by the client to indicate that it has
+  finished sending requests
+
+The requests are stored in a buffer until they can be sent; the buffer capacity can
+be specified in the `BidirectionalStreamingSettings` passed into the initial call.
+
+The difference between the methods with a `Try` prefix and those without is about error
+handling:
+
+- A call to `WriteAsync` will throw an exception if the client stream has already been
+  completed, or if the buffer is full. A call to `TryWriteAsync` in the same failure
+  scenarios will return a null `Task` to indicate failure instead.
+- A call to `CompleteAsync` will throw an exception if the client stream has already been
+  completed. A call to `TryCompleteAsync` will return a null `Task` to indicate failure
+  instead.
+
+In most cases, it's more appropriate to use `WriteAsync` and `CompleteAsync`; the `Try` variants
+are provided for situations where you may have multiple threads sending requests and potentially
+completing the stream.
+
+Unlike the "raw" gRPC stub, users do not need to ensure that only a single request is in-flight
+at a time: the buffer takes care of handling multiple requests, so long as the buffer capacity
+is not exceeded. The tasks returned by the above methods complete when the request has been
+written to the underlying stream (or when the stream has actually been marked as completed,
+for `TryCompleteAsync` and `CompleteAsync`).
```

Comment:
If the "or" part happens because on completion the buffer is flushed, the it might make sense to say so:
```suggestion
by `TryCompleteAsync` and `CompleteAsync`, because the buffer is flushed and pending requests are written to the underlying stream).
```

Reply:
Ah wait, you did mean for, as in the tasks for the Complete methods complete when....

Ah but it is confusing, because you are talking about two distinct types of tasks. Maybe, talk about them separately in this whole paragrpah:

"The tasks returned by the write methods complete when the request has been written to the underlying stream and not the buffer.
The tasks returned by the complete methods complete when all pending buffered requests have been written to the underlying stream and it has been marked as completed."

Reply:
Yup - I've used a slightly different formatting, but separated them out. I've also added another sentence to make it clear that completion only happens when the buffer is empty.